submission/proxy_server.go

```diff
@@ -45,6 +46,20 @@ func NewProxyServer(logListPath string, dBuilder DistributorBuilder, reqTimeout
 // Run starts regular Log list updates. Blocks until initialization happens.
 func (s *ProxyServer) Run(logListRefreshInterval time.Duration, rootsRefreshInterval time.Duration) {
 	s.p.Run(context.Background(), logListRefreshInterval, rootsRefreshInterval)
+
+	// start listening to proxy errs
+	go func() {
+		for {
+			select {
+			case <-context.Background().Done():
+				return
+			case err := <-s.p.Errors:
```

Comment:
What is supposed to happen in the case of an error?

Reply:
Just logging. States that require manual changes should be flagged via monitoring tools.

Reply:
Why don't we just log these errors where they occur?

Reply:
As we want to unify their logging. That's intermediate collection point.

Reply:
Why do we want that? This seems like an antipattern to me.

By posting errors to a channel you enforce the `Proxy` type clients to read from this channel, otherwise it will block on writing to it on the second error (since the channel size is 1). This behavior of `Proxy` type is not obvious and should be documented.

Your comment below regarding both critical and non-critical errors going to this channel is another reason to log errors where they occur, because those are the places where there is more context on whether it's a `Warning` or a more severe `Error`.

Finally, log messages have filenames and line numbers in them. If you log them all from the same place then this information becomes useless (all messages come from the same file and line), which makes: a) diagnosing problems more difficult; b) [V-logging](https://godoc.org/github.com/golang/glog#V) less useful.